cpp/src/arrow/engine/substrait/serde_test.cc

```diff
@@ -2082,43 +2103,121 @@ TEST(Substrait, BasicPlanRoundTrippingEndToEnd) {
+      [5, 3, 12],
+      [1, 3, 12]
+    ])"});
+
+  auto format = std::make_shared<arrow::dataset::IpcFileFormat>();
+  auto filesystem = std::make_shared<fs::LocalFileSystem>();
+  const std::string file_name = "serde_project_test.arrow";
+
+  ASSERT_OK_AND_ASSIGN(auto tempdir,
+                       arrow::internal::TemporaryDir::Make("substrait-tempdir-project-"));
+  ASSERT_OK_AND_ASSIGN(auto file_path, tempdir->path().Join(file_name));
+  std::string file_path_str = file_path.ToString();
+
+  WriteIpcData(file_path_str, filesystem, table);
+
+  std::vector<fs::FileInfo> files;
+  const std::vector<std::string> f_paths = {file_path_str};
+
+  for (const auto& f_path : f_paths) {
+    ASSERT_OK_AND_ASSIGN(auto f_file, filesystem->GetFileInfo(f_path));
+    files.push_back(std::move(f_file));
   }
-  ASSERT_OK_AND_ASSIGN(auto rnd_trp_table,
-                       GetTableFromPlan(roundtripped_filter, exec_context, dummy_schema));
-  EXPECT_TRUE(expected_table->Equals(*rnd_trp_table));
+
+  ASSERT_OK_AND_ASSIGN(auto ds_factory, dataset::FileSystemDatasetFactory::Make(
+                                            filesystem, std::move(files), format, {}));
+  ASSERT_OK_AND_ASSIGN(auto dataset, ds_factory->Finish(dummy_schema));
+
+  auto scan_options = std::make_shared<dataset::ScanOptions>();
+  scan_options->projection = compute::project({}, {});
+  compute::Expression project_fp_expr = compute::call(
+      "add", {compute::field_ref(FieldPath({1})), compute::field_ref(FieldPath({2}))});
+
+  std::vector<compute::Expression> project_fp_exprs = {project_fp_expr};
+
+  auto declarations = compute::Declaration::Sequence(
+      {compute::Declaration(
+           {"scan", dataset::ScanNodeOptions{dataset, scan_options}, "s"}),
+       compute::Declaration(
+           {"project", compute::ProjectNodeOptions{project_fp_exprs}, "p"})});
+
+  // adding the project expression field to schema
+  auto project_schema = schema({field("add(shared,distinct)", int32())});
+
+  ASSERT_OK_AND_ASSIGN(auto expected_table,
+                       GetTableFromPlan(declarations, exec_context, project_schema));
+
+  std::shared_ptr<ExtensionIdRegistry> sp_ext_id_reg = MakeExtensionIdRegistry();
+  ExtensionIdRegistry* ext_id_reg = sp_ext_id_reg.get();
+  ExtensionSet ext_set(ext_id_reg);
+
+  ASSERT_OK_AND_ASSIGN(auto serialized_plan, SerializePlan(declarations, &ext_set));
+
+  ASSERT_OK_AND_ASSIGN(
+      auto sink_decls,
+      DeserializePlans(
+          *serialized_plan, [] { return kNullConsumer; }, ext_id_reg, &ext_set));
+  // assert project declaration
+  const auto& roundtripped_emit_project =
+      std::get<compute::Declaration>(sink_decls[0].inputs[0]);
+
+  // assert emit projection declaration
+  // emit node used in the serialization outputs the 4th column since the
+  // 0, 1, 2 original columns from the input data and 4th column being the
+  // projected column.
+  auto expec_project_expr = {compute::field_ref(FieldPath({4}))};
```

Comment:
This is also doubtful.

Reply:
Right now I think our Substrait->Acero path does not special case the "project" relation when handling emit.  In other words, converting Substrait->Acero could convert one project node into two back-to-back project nodes.  So I don't know that project will ever round trip.  What I would expect in this case (I think) is two project nodes.

First project node (four expressions):

field(0), field(1), field(2), field(1) + field(2)

Second project node (one expression):

field(3)

I'm not sure why it is `4` though instead of `3`.